include/pagmo/problems/unconstrain.hpp

```diff
@@ -0,0 +1,379 @@
+     * and \p T is not pagmo::problem.
+     *
+     * Wraps a user-defined problem so that its constraints will be removed
+     *
+     * @param p a user-defined problem.
+     * @param method an <tt>std::string</tt> containing the name of the method to be used t remove the constraints: one
+     * of "death penalty", "kuri", "weighted", "ignore_c" or "ignore_o".
+     * @param weights an <tt>std::vector</tt> containing the weights in case "weighted" is selected as method.
+     *
+     * @throws std::invalid_argument if the length of \p weights is
+     * not equal to the problem constraint dimension \f$ n_{ec} + n_{ic}\f$ when \p method is "weighted", if the
+     * \p method is not one of "death penalty", "kuri", "weighted", "ignore_c" or "ignore_o", if the \p weights vector
+     * is not empty and the \p method is not "weighted" or if \p is already unconstrained
+     * @throws unspecified any exception thrown by the pagmo::problem constructor
+     */
+    template <typename T, ctor_enabler<T> = 0>
+    explicit unconstrain(T &&p, const std::string &method, const vector_double &weights = vector_double{})
+        : problem(std::forward<T>(p)), m_weights(weights)
+    {
+        // The number of constraints in the original udp
+        m_nec = static_cast<const problem *>(this)->get_nec();
+        m_nic = static_cast<const problem *>(this)->get_nic();
+        m_nc = m_nec + m_nic;
+        // 1 - We throw if the original problem is unconstrained
+        if (m_nc == 0u) {
+            pagmo_throw(std::invalid_argument, "Unconstrain can only be applied to constrained problems");
+        }
+        // 2 - We throw if the method weighted is selected but the weight vector has the wrong size
+        if (weights.size() != m_nc && method == "weighted") {
+            pagmo_throw(std::invalid_argument, "Length of weight vector is: " + std::to_string(weights.size())
+                                                   + " while the problem constraints are: " + std::to_string(m_nc));
+        }
+        // 3 - We throw if the method selected is not supported
+        if (method != "death penalty" && method != "kuri" && method != "weighted" && method != "ignore_c"
+            && method != "ignore_o") {
+            pagmo_throw(std::invalid_argument, "The method " + method + " is not supported (did you mispell?)");
+        }
+        // 4 - We throw if a non empty weight vector is passed but the method weghted is not selected
+        if (weights.size() != 0u && method != "weighted") {
+            pagmo_throw(std::invalid_argument,
+                        "The weight vector needs to be empty to use the unconstrain method " + method);
+        }
+        // 5 - We store the method in a more efficient enum type and the number of objectives of the orginal udp
+        std::map<std::string, method_type> my_map = {{"death penalty", method_type::DEATH},
```

Comment:
Since this map is used in at least one other place, maybe it would be better to have it stored as a static const variable, perhaps with the usual template trick.